Resend messages on failure

ramhoj opened the pull request:
closes #29
contributes to #4
suggests that #3 dependency injection might be a pragmatic first step to test more of the code*

This, quite large PR, makes the lib resend messages that comes after an error response from Apple since these messages are otherwise never delivered.

There are a few things that can be improved that we might add in a later PR that is related to re-sending messages that this PR doesn't include:
- re-try on :ssl error
- handle message expiry

The queue implementation is inspired by apns4erl.

To verify that this addition actually works we found our self in a position where we really wanted a better safety net in the form of unit tests to rely on. Therefore this PR also includes a split of `APNS.Worker` into several smaller modules and adds unit tests for most of them.

Please let us know what you think, we're happy to get feedback on naming, separation etc.

\* We tried using https://github.com/pdincau/mockapn which caused a lot of trouble (we couldn't find a way to start and stop the service in a dependable way)

Linuus committed:
Adds basic support for resending notifications on fail

Commit:
Simplify cert setup when contributing

Commit:
Split worker into modules and add unit tests

zepplock commented:
This is pretty big. How well was it tested in your environment?

ramhoj commented:
@zepplock we've tested it with our server backend that has apns4ex as a dependency on our staging environment that uses Apple's sandbox environment and have confirmed that sending push notifications still work (they are received by the device).

When we tried to send to a valid token after an invalid the valid token still received.

Have you had time to have a look at the diff? Anything that stands out? Feedback?

Commit:
Cast message ids to integer-32 everywhere

This is needed to be able to compare messages
in the queue so that they aren't sent twice.

zepplock commented:
I did review the diffs and they looked great. Let's do some testing on it. So proposed plan would be:
- you test in your env
- I will get code, compile, run on my server. I don't push a lot of traffic with it but at least will have it deployed to staging server.
- If testing goes fine I will bump version to 1.0.0 and start doing proper SEMVER for it (unless @chvanikoff objects)

It would be great if somebody else reviews the code.

benwilson512 commented on lib/apns/configuration.ex:
```diff
@@ -0,0 +1,102 @@
+    ssl_opts = [reuse_sessions: false, mode: :binary]
+
+    if config.certfile != nil do
+      ssl_opts = Dict.put(ssl_opts, :certfile, certfile_path(config.certfile))
```
Why are you using Dict.put? You define the dict on line 55, it's a keyword list.

benwilson512 commented on lib/apns/configuration.ex:
```diff
@@ -0,0 +1,102 @@
+
+    if config.key != nil do
+      case :public_key.pem_decode(config.key) do
+        [{:RSAPrivateKey, keyDer, _}] -> ssl_opts = Dict.put(ssl_opts, :key, {:RSAPrivateKey, keyDer})
```
You should refactor these clauses to avoid imperative style assignment inside conditionals. The option building code in general could probably be factored into its own function, this function has gotten rather long.

benwilson512 commented on lib/apns/configuration.ex:
```diff
@@ -0,0 +1,102 @@
+      ssl_opts = Dict.put(ssl_opts, :keyfile, Path.absname(config.keyfile))
+    end
+
+    {:ok, queue_pid} = APNS.Queue.start_link
```
These needs to be started inside of a supervisor. Without starting it in a supervisor it won't be shutdown properly, and hot code reloading is impossible.

ramhoj replied:
@benwilson512 We're not quite sure what the right way to supervise the queue here is since we're (at least we think we do) want the queue to stop when the worker it is spawned in is stopped.

If we make a global supervisor that starts with the app and then adds the queue process to that supervisor with `start_child` what happens then when the worker has finished and is closed?

Could you provide an example of what you think is the proper way?

benwilson512 replied:
One question is why the queue is its own process to begin with. It looks like it's only ever used from inside another genserver, why can't the genserver maintain its own queue in the process state?

benwilson512 commented on lib/apns/payload.ex:
```diff
@@ -0,0 +1,121 @@
+      payload = put_in(payload[:aps][:alert], message.alert)
+    else
+      payload = put_in(payload[:aps][:alert], format_loc(message.alert))
+    end
```
Refactor to avoid imperative cases. Consider trying to use more pattern matching.

benwilson512 commented on lib/apns/payload.ex:
```diff
@@ -0,0 +1,121 @@
+    if launch_image != nil do
+      alert = alert
+      |> Map.put(:'launch-image', launch_image)
+    end
```
Refactor to avoid imperative cases. This pattern can be simplified by having a function that takes a list of values, and a list of defaults. It then iterates through whatever the dict is and if a given value is nil, sets the default.

benwilson512 commented on lib/apns/worker.ex:
```diff
@@ -2,336 +2,64 @@ defmodule APNS.Worker do
-
-    encode(payload, payload_limit)
+  defp sleep(seconds) do
+    :timer.sleep(seconds * 1000)
```
Why are you sleeping in a genserver?

ramhoj replied:
@zepplock, @chvanikoff what do you say? We just moved this from the original code into this file.

benwilson512 commented on lib/apns/worker.ex:
```diff
@@ -2,336 +2,64 @@ defmodule APNS.Worker do
-    }
-    Map.merge(config, config2)
+  defp send_after(seconds, message) do
+    :erlang.send_after(seconds * 1000, self, message)
```
`Process.send_after` is considered more idiomatic

ramhoj commented:
Sounds great @zepplock. We addressed most of @benwilson512's feedback and updated the PR. We have now run our branch in our staging environment for about 2 days and we feel that we're ready to try it in our production environment tomorrow (9am-ish GMT+1).

If we encounter no bugs there for a couple of days it should be a decent indication that we haven't introduced any obvious bugs.

Regarding 1.0.0 we would perhaps want to fix #33 before it's release as we would sort of break the API (the documentation is right, but if someone has side stepped the bug by setting a epoch timestamp them self, they would get a different result after #33 is fixed).

benwilson512 commented on lib/apns/state.ex:
```diff
@@ -0,0 +1,16 @@
+defmodule APNS.State do
+  def get(options) do
+    {:ok, queue_pid} = APNS.Queue.start_link
```
Moving the start link discussion here because the other one is on an outdated diff. Basicaly, why is the queue a separate process? It lives and dies with the main genserver, and you could just have `queue: []` down on line 12 and do just as well.

benwilson512 replied:
Better really because you won't have message passing overhead, we don't have to solve the problem of having the queue supervised, etc.

Linuus replied:
That is true. I guess we can just move it into the worker genserver. Thanks!

Commit:
Move queue into the worker process state

Linuus commented:
@benwilson512 @zepplock We moved the queue stuff into the worker process as @benwilson512 suggested. Please have another look :+1:

We also have some pending tests we'd appreciate some feedback on:
https://github.com/chvanikoff/apns4ex/pull/32/files#diff-8d90ecd04c79988a3c0d8b4c15f5a8efR207
https://github.com/chvanikoff/apns4ex/pull/32/files#diff-8d90ecd04c79988a3c0d8b4c15f5a8efR219

Finally (this should be passing?):
https://github.com/chvanikoff/apns4ex/pull/32/files#diff-8d90ecd04c79988a3c0d8b4c15f5a8efR77

benwilson512 commented on lib/apns/message_handler.ex:
```diff
@@ -0,0 +1,82 @@
+  end
+
+  defp messages_after(queue, failed_id) do
+    Enum.take_while(queue, fn(message) -> message.id != failed_id end)
```
So, this is treating the queue like a stack instead of a queue. That may be what you want, but that seems doubtful. New messages are accumulated at the head of the list, so suppose you have a queue that looks like `[m4, m3, m2, m1]`. Presumably you want m1 to be handled before m2 yes? Your take_while logic is still sound, you just need to reverse the result so that when you run it through the list comprehension on line 63 it does them in the right order.

``` elixir
queue
|> Enum.take_while(fn(message) -> message.id != failed_id end)
|> Enum.reverse
```

Ordinarily I'd say use the `:queue` module but given that you just discard whatever remains in the queue online 67 that seems unnecessary.

Linuus replied:
@benwilson512 Yes, it's returned in reverse but it doesn't really matter, as we just want to resend all the messages. But sure, it might be more intuitive to reverse first.

Yes, I looked at the erlang `:queue` module first, but it seemed a bit overkill for our use case.

benwilson512 replied:
Resending the messages in the reverse order isn't a problem? I mean that's fine, I would just explicitly note that in a comment or something.

Linuus replied:
Actually now that I think more about it, it doesn't matter in our use case but it may be weird for other use cases... We'll just reverse it! Thanks :+1:

benwilson512 commented on lib/apns/message_handler.ex:
```diff
@@ -0,0 +1,82 @@
+        APNS.Error.new(message_id, status) |> state.config.callback_module.error()
+
+        for message <- messages_after(state.queue, message_id) do
+          GenServer.cast(worker_pid, message)
```
So, why do this genserver cast all of these messages to itself instead of just processing them inline?

Linuus replied:
Why not? Isn't it good practice to do that instead of calling `handle_cast` directly? That way it can process other messages as well, right? If we call `handle_cast` directly we'll be stuck here until all the notifications has been sent instead of just queueing them up and move on.

benwilson512 replied:
well there's sort of 2 issues here. 1 is that because the message is opaque, I it wasn't clear what this was actually doing. In the end I could figure it out because the worker only happens to have one `handle_cast` function, but in general this is why there are wrapping functions in genserver modules. This should really be

in https://github.com/elabs/apns4ex/blob/resend-messages-on-failure/lib/apns/worker.ex

``` elixir
def push(worker_pid, message) do
  GenServer.cast(worker_pid, {:push, message})
end

def handle_cast({:push, message}, state), do: etc...
```

that way when you use it here you can do `Worker.push(worker_pid, message)` and readers have a better clue about what's going on.

I do see your point now, yes it's nice to handle one message at a time. Basically the process's message queue becomes the real queue. I feel like there's a way to do this that uses the process state queue you already have, but this seems fine.

benwilson512 replied:
I should add that `push` doesn't quite communicate `resend` which is basically what this is.

Linuus replied:
Yes, I totally agree that we probably shouldn't do `GenServer.cast` here but create a better public API in the worker.

We'll figure something out.

ramhoj replied:
TL;DR: what about coupling?

I fully agree about the readability and it's being difficult to track matching `handle_cast` functions but don't we couple `MessageHandler` and `Worker` even tighter together if we invoke `Worker.push(worker_pid, message)` from the `MessageHandler`? When [passing in a pid](https://github.com/chvanikoff/apns4ex/pull/32/files#diff-888c2a7eb39a7c40aa4f62080c8e5e22R58) the recipient isn't known to the caller which should be a good thing, right?

Of course we could send in the Worker-module as well, but at that point it feels a bit convoluted.

To me the most readable is to not call back to `Worker` at all on re-tries (just call `push(message, state)` [here](https://github.com/chvanikoff/apns4ex/pull/32/files#diff-888c2a7eb39a7c40aa4f62080c8e5e22R64) but as @Linuus mentioned that would delay return in `MessageHanlder.handle_respone/4`.

ramhoj replied:
TL;DR: separate worker in to one for feedback and one for messages.

After discussing this more with @Linuus we've started to think that we're attacking this from the wrong angle. Before this PR both `MessageHandler`, `FeedbackHandler`, and `Worker` was in the same file. We wanted to protect against regressions and therefore added unit tests for this code. We didn't want to test all the logic through the GenServer interface so we basically just copy-pasted business logic into its own modules (`MessageHandler` and `FeedbackHandler`).

Perhaps this was wrong, what if we instead make two different Workers, one for Feedback and one for Message? Then we can inline the business logic in the Workers which would improve readability and not create falsey illusions that the Handlers are properly separated from the Worker.

This would also allow us to configure number of workers individually for Messages and Feedback. From what we understand now, if we want, say 100, worker because we want to send a lot of messages on separate connections we have the side-effect that we'll also call the Feedback-service 100 times each interval?

We haven't started on this yet and there might be some good arguments why this is a bad idea, so we're just putting it out there to hear what you think.

Commit:
Log too long tokens in error callback

Commit:
Send message back to poolboy pool on ssl error

Commit:
Calculate expiry as offset from epoch offset

Commit:
Calculate expiry based on when message was first generated

Commit:
Limit the number of ssl send retries

Commit:
Reply when closing worker

Commit:
Use Connection API to manage closed ssl connections

Commit:
Rename worker module

Commit:
Inline FeedbackHandler in FeedbackWorker

Commit:
Inline MessageHandler in MessageWorker

Commit:
Remove reconnect after x messages feature

Commit:
Fix bad public function call

Commit:
Use cast in Message worker

Commit:
Remove unneeded Logger require

Commit:
Don't use overflow workers

ramhoj commented:
@zepplock after some more testing in a prod env we did experience regression. We started to get a lot of :ssl errors (connections closed). We've updated the the PR to use the [Connection library](https://github.com/fishcakez/connection) to maintain a healthy connection, thus allowing this PR to close #32 

This also made us split the feedback and message sending into separate workers as discussed in earlier comment.

We wrote a script to stress-test the implementation by sending x invalid tokens followed by a valid and repeating that for about 5 times (https://gist.github.com/ramhoj/0eba365d86670f9f1933c8e6ce1e6603)

Unfortunately the result wasn't prefect. The two devices that we owned with valid tokens did not always receive the push with this setup. The higher number of invalid pushes sent before the valid once seemed to lower the delivery rate. Except for the known cases which we now handles no errors where raised so it's hard to tell where it goes wrong. Our guess is that somehow Apples feedback service don't like such high failure-rate but perhaps this is just wishful thinking.

The new changes introduces a new `APNS.push_parallel` function as discussed in #36 but with no chunking. This could be a problem if called millions of times as there will be an equal number of processes spawned.

These changes also closes #33 with the disclaimer that a low max delivery time can't be guaranteed if it takes the lib longer to deliver the message to apple than the configured expiry time.

zepplock commented:
My take on this is that it's an awesome work and adds a lot of value for the library. I'm thinking of making this version `0.9` and let people try it. This way people can continue using the old one without breaking things. If we hear positive reviews we will bump this to `1.0`. What do you guys think?

ramhoj commented:
Thanks @zepplock! Sounds smart, I'd mark 0.9 as some kind of beta in the readme and inform people that they can use the previous version without the resending-fix as the stable one until we've had time to test more.

Commit:
Review log levels

Commit:
Log token of payload exceeded error

Commit:
Better to use GenServer message box than process

We get unstable results on high volumes when using
push_parallel.

Commit:
Make token available in error callback

Linuus commented:
@zepplock Sounds good.

We just pushed a few changes.

**Removed `push_parallel`**
It used to wrap the call to `push` in a `Task.start` to make it async. Since we use a cast now for the push messages it shouldn't be necessary anymore as it returns very quickly. When we tested it locally it seemed to be using less resources.

**Pass token to error callback**
When Apple returns an error it only returns the message id but since we have the message in queue we can find the token there and pass it to the error callback. This is useful if you want to log the failed message token. Also, if you get the `Invalid token` error (8) you may want to handle this in the eror callback (delete probably since it's invalid).

zepplock commented:
I will test this on my setup and push it out.

Commit:
Make explicit case for empty queue when fetching token in handle_response
